doc/check-log4j.1
.Dd February 28, 2022
.Dt check-log4j 1
.Os
.Sh NAME
.Nm check-log4j
.Nd try to determine if a host is vulnerable to log4j CVE-2021-44228
.Sh SYNOPSIS
.Nm
.Op Fl Vhv
.Op Fl j Ar jar
.Op Fl p Ar path
.Op Fl s Ar skip
.Sh DESCRIPTION
The
.Nm
tool attempts to determine whether the host it is
executed on is vulnerable to the log4j RCE
vulnerability identified as CVE-2021-4428.
.Pp
Since this vulnerability is in a specific Java class
that may be inside nested Java archive files,
.Nm
may be somewhat intrusive to run and should be
executed with care and consideration of the system's
load.
Please see DETAILS for more information.
.Sh OPTIONS
The following options are supported by
.Nm :
.Bl -tag -width p_path_
.It Fl V
Print version number and exit.
.It Fl h
Print a short help message and exit.
.It Fl j Ar jar
Check only this archive, nothing else.
Can be specified multiple times for multiple JAR
(or other zip formatted archive) files.
.It Fl p Ar path
Limit filesystem traversal to this directory.
Can be specified multiple times.
If not specified,
.Nm
will default to '/'.
.It Fl s Ar skip
Skip the given checks.
Valid arguments are 'files', 'packages', and
\'processes'.
.It Fl v
Be verbose.
Can be specified multiple times.
.El
.Sh DETAILS
CVE-2021-4428 describes a possible remote code
execution (RCE) vulnerability in the popular log4j
framework.
Simply causing the vulnerable system to log a
specifically crafted message can the attacker gain
command execution and information disclosure
capabilities.
This vulnerability relies on an insecure default
setting applying to the Java Naming and Directory
Interface (JNDI).
.Pp
Specifically, a system that contains the
JndiLookup.class may enable the attack path in
question.
.Pp
To determine whether a host is vulnerable, the
.Nm
tool will perform the following checks:
.Bl -bullet -compact
.It
check for the existence of likely vulnerable packages
.It
check for the existence of java processes using the
\'JndiLookup' class
.El
.Pp
The discovery process may include running
.Xr find 1 ,
.Xr lsof 1 ,
or
.Xr rpm 1 ;
please use the
.Fl s
flag to skip any checks that might have a negative
impact on your host.
.Pp
The output of the command attempts to be human
readable and provide sufficient information to judge
whether the host requires attention.
.Sh ENVIRONMENT
The following environment variables influence the
behavior of
.Nm :
.Bl -tag
.It CHECK_LOG4J_FIND_OPTS_PRE
Additional options to pass to
.Xr find 1
prior to the path name(s).
.Pp
By default,
.Nm
runs "find / -type f -name '*.[ejw]ar'";
the contents of this variable are placed immediately
after the 'find' and before the path name(s).
.It CHECK_LOG4J_FIND_OPTS_POST
Additional options to pass to
.Xr find 1
immediately after the path name(s).
.El
.Sh EXAMPLES
Sample invocation on a non-vulnerable host:
.Bd -literal -offset indent
$ check-log4j
No obvious indicators of vulnerability found.
$
.Ed
.Pp
Sample invocation only looking at processes
.Bd -literal -offset indent
$ ./check-log4j.sh -s files -s packages -v -v
=> Running all checks...
==> Skipping package check.
==> Looking for jars...
==> Skipping files check.
==> Checking all found jars...
check-log4j.sh 1.0 localhost: Possibly vulnerable jar 'BOOT-INF/lib/log4j-core-2.14.1.jar' (inside of /home/jans/log4shell-vulnerable-app-0.0.1-SNAPSHOT.jar) used by process 15569.

$
.Ed
.Pp
Sample invocation searching only /var and /usr/local/lib
and skipping package and process checks:
.Bd -literal -offset indent
$ check-log4j -p /var -p /usr/local/lib -s packages -s processes
Possibly vulnerable jar '/usr/local/lib/jars/log4j-core-2.15.0.jar'.
Possibly vulnerable jar '/usr/local/lib/jars/log4j-core-2.15.jar'.
Possibly vulnerable jar '/usr/local/lib/jars/log4j-core-2.jar'.
Possibly vulnerable jar '/usr/local/lib/jars/log4j-core.jar'.

$
.Ed
.Pp
Note version comparisons are only done for packages,
which is why the above output incudes files ending in
a seemingly non-vulnerable version.
.Pp
To avoid mountpoint traversal on a Unix system where
.Xr find 1
requires the
.Fl x
flag to precede the paths:
.Bd -literal -offset indent
$ env CHECK_LOG4J_FIND_OPTS_PRE="-x" check-log4j
No obvious indicators of vulnerability found.
.Ed
.Pp
To only search files newer than '/tmp/foo':
.Bd -literal -offset indent
$ env CHECK_LOG4J_FIND_OPTS_POST="-newer /tmp/foo" check-log4j
No obvious indicators of vulnerability found.
.Ed
.Sh EXIT STATUS
.Nm
will return 0 if the host was found not to be
vulnerable and not in need of any update;
it will return 1 if a vulnerable jar or package was
detected.
.Pp
If no vulnerability to CVE-2021-44228 / CVE-2021-45046
was found, but versions below the desired minimum were
found,
.Nm
will return 2.
.Sh SEE ALSO
.Xr find 1 ,
.Xr lsof 1 ,
.Xr rpm 1
.Sh HISTORY
.Nm
was originally written by
.An Jan Schaumann
.Aq jans@yahooinc.com
in December 2021.
.Sh BUGS
Please file bugs and feature requests via GitHub pull
requests and issues or by emailing the author.


CHANGES
Version 3.0 (2022-02-28)
 * remove '-f' option

Version 2.4 (2022-02-28)
 * handle corrupt zip files more gracefully

Version 2.3 (2022-01-27)
 * bug fix to really match all pathnames case-insensitively

Version 2.2 (2022-01-26)
 * match all pathnames case-insensitively

Version 2.1 (2022-01-11)
 * suppress warnings from lsof(8)
 * handle zero-sized files more gracefully
 * perform case-insensitive search
 * use 'command -v' instead of which(1)

Version 2.0 (2022-01-03)
 * still note if we encounter versions <= 2.17.1

Version 1.9 (2021-12-17)
 * use correct IFS

Version 1.8 (2021-12-17)
 * add '-V' to report version number

Version 1.7 (2021-12-17)
 * allow user to pass flags to find(1)
 * also check EAR files

Version 1.6 (2021-12-16)
 * correctly detect fixed versions
 * use full directory names to ensure find(1) traverses symlinks

Version 1.5 (2021-12-16)
 * also look into WAR files

Version 1.4 (2021-12-15)
 * detect fatal class even if not inside a jar
 * first open source version

Version 1.3 (2021-12-15)
 * better 2.16.0 detection
 * handle relative directories better
 * use more portable ps(1) flags
 * handle given jar files better

Version 1.2 (2021-12-15)
 * logic error

Version 1.1 (2021-12-15)
 * if unzip(1) isn't found, try grep(1)

Version 1.0 (2021-12‐15)
 * first version number introduced
 * rapid iterations and changes from the initial
   version pushed to git on 2021-12-13


Version 0.1 (2021-12-13)
 * hello ${jndi:ldap://world/!}


src/check-log4j.sh
#! /bin/sh
#
# Originally written by Jan Schaumann
# <jans@yahooinc.com> in December 2021.
#
# This script attempts to determine whether the host
# it runs on is likely to be vulnerable to log4j RCE
# CVE-2021-44228 / CVE-2021-45046.
#
# Copyright 2021 Yahoo Inc.
# 
# Licensed under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in
# compliance with the License.  You may obtain a copy of
# the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in
# writing, software distributed under the License is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing
# permissions and limitations under the License.


set -eu
IFS="$(printf '\n\t')"

umask 077

###
### Globals
###

PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin

# We broadly only care about versions >= 2.16.
# 1.x has been determined not to be vulnerable, and we
# boldly hope any version after 2.16 (including new
# major versions, if any) will not regress.
MAJOR_WANTED="2"
MINOR_MINIMUM="16"

# CVE-2021-44832 affects log4j <= 2.17.0, but requires
# the attacker to control the config change, so really
# not critical; we still _want_ folks to update,
# though.
MINOR_WANTED="17"
TINY_WANTED="1"

# log4j-2.16.0 _disables_ JNDI lookups, but leaves in
# place the class, meaning it could be enabled if
# log4j2.enableJndi=true.
KNOWN_DISABLED="log4j-core-${MAJOR_WANTED}.${MINOR_MINIMUM}"
FATAL_SETTING="-Dlog4j2.enableJndi=true"
FATAL_CLASS="JndiLookup.class"

_TMPDIR=""
CHECK_JARS=""
ENV_VAR_SET="no"
FOUND_JARS=""
PROGNAME="${0##*/}"
RETVAL=1
SEARCH_PATHS=""
SKIP=""
SEEN_JARS=""
SHOULD_UPGRADE="unknown"
SUSPECT_CLASSES=""
SUSPECT_JARS=""
SUSPECT_PACKAGES=""
UNZIP="$(command -v unzip 2>/dev/null || true)"
VERBOSITY=0
VERSION="3.0"

LOGPREFIX="${PROGNAME} ${VERSION} ${HOSTNAME:-"localhost"}"

###
### Functions
###


cdtmp() {
	if [ -z "${_TMPDIR}" ]; then
		_TMPDIR=$(mktemp -d ${TMPDIR:-/tmp}/${PROGNAME}.XXXX)
	fi
	cd "${_TMPDIR}"
}

checkFilesystem() {
	local class=""
	local classes=""
	local okVersion=""
	local newjars=""
	local findCmd=""

	if expr "${SKIP}" : ".*files" >/dev/null; then
		verbose "Skipping files check." 2
		return
	fi

	verbose "Searching for java archives on the filesystem..." 3
	findCmd=$(echo find ${CHECK_LOG4J_FIND_OPTS_PRE:-""} "${SEARCH_PATHS:-/}" ${CHECK_LOG4J_FIND_OPTS_POST:-""})

	verbose "Running '${findCmd}'..." 4

	newjars=$(eval ${findCmd} -type f -iname \'*.[ejw]ar\' 2>/dev/null || true)
	FOUND_JARS="$(printf "${FOUND_JARS:+${FOUND_JARS}\n}${newjars}")"

	verbose "Searching for ${FATAL_CLASS} on the filesystem..." 3
	classes=$(eval ${findCmd} -type f -iname "${FATAL_CLASS}" 2>/dev/null || true)

	for class in ${classes}; do
		okVersion="$(checkFixedVersion "${class}")"
		if [ -z "${okVersion}" ]; then
			log "Possibly vulnerable class ${class}."
			SUSPECT_CLASSES="$(printf "${SUSPECT_CLASSES:+${SUSPECT_CLASSES}\n}${class}")"
		fi
	done
}

checkFixedVersion() {
	local file="${1}"
	local ver=""
	local mgrClass=""
	local suffix="${file##*.}"
	local dir=""

	set +e
	if [ x"${suffix##*[ejw]}" = x"ar" ]; then
		if [ -z "${UNZIP}" ]; then
			warn "Unable to check if ${suffix} contains a fixed version since unzip(1) is missing."
			return
		fi
		verbose "Checking for fixed classes in '${file}'..." 6
		if zeroSize "${file}"; then
			verbose "Skipping zero-size file '${file}'..." 6
			return
		fi

		mgrClass="$(${UNZIP} -q -l "${file}" 2>/dev/null | awk 'tolower($0) ~ /jndimanager.class$/ { print $NF; }')"
	if [ -n "${mgrClass}" ]; then
		cdtmp
			${UNZIP} -o -q "${file}" "${mgrClass}" 2>/dev/null
		fi
	elif [ x"${suffix}" = x"class" ]; then
		# If we find the fatal class outside of a jar, let's guess that
		# there might be an accompanying JndiManager.class nearby...
		mgrClass="${file%/*}/../net/JndiManager.class"
	fi

		if [ -f "${mgrClass}" ]; then
			if grep -q 'log4j2.enableJndi' "${mgrClass}" ; then
				echo "log4j2.enableJndi found"
			fi
		fi
	set -e
}

checkInJar() {
	local jar="${1}"
	local needle="${2}"
	local pid="${3}"
	local parent="${4:-""}"
	local msg=""
	local match=""
	local flags=""
	local okVersion=""
	local rval=0

	local thisJar="${parent:+${parent}:}${jar}"
	for j in $(echo "${SEEN_JARS}" | tr ' ' '\n'); do
		if [ x"${j}" = x"${thisJar}" ]; then
			verbose "Skipping already seen archive '${thisJar}'..." 6
			return
		fi
	done
	SEEN_JARS="${SEEN_JARS:+${SEEN_JARS} }${thisJar}"

	if [ -z "${parent}" ]; then
		if zeroSize "${thisJar}"; then
			verbose "Skipping zero-size file '${thisJar}'..." 6
			return
		fi
	fi

	verbose "Checking for '${needle}' inside of ${jar}..." 5

	set +e
	if [ -n "${UNZIP}" ]; then
		${UNZIP} -q -l "${jar}" 2>/dev/null | grep -q "${needle}"
	else
		warn "unzip(1) not found, trying to grep..."
		grep -q "${needle}" "${jar}"
	fi
	rval=$?
	set -e

	if [ ${rval} -eq 0 ]; then
		if [ -n "${parent}" ]; then
			msg=" (inside of ${parent})"
		fi
		if [ x"${jar}" != x"${pid}" ] && expr "${pid}" : "[0-9]*$" >/dev/null; then
			if checkPid "${pid}" ; then
				flags="JNDI Lookups enabled via command-line flags"
			fi
			msg="${msg} used by process ${pid}"
		fi

		okVersion="$(checkFixedVersion "${jar}")"

		# We're specifically looking for a jar, so no need to match .[ew]ar # here.
		match="$(echo "${jar}" | sed -n -e "s|.*/\(${KNOWN_DISABLED}[0-9.]*.jar\)$|\1|p")"
		if [ -n "${match}" -o -n "${okVersion}" ]; then
			if [ -n "${flags}" ]; then
				log "Normally non-vulnerable archive '${jar}'${msg} found, but ${flags}!"
			fi
			verbose "Allowing archive '${jar}' with known disabled JNDI Lookup." 6
			return
		fi
		if [ -z "${flags}" ]; then
			log "Possibly vulnerable archive '${jar}'${msg}."
		fi
		SUSPECT_JARS="${SUSPECT_JARS} ${thisJar}"
	fi
}

checkJars() {
	local found jar jarjar msg pid

	if [ -z "${CHECK_JARS}" ]; then
		findJars
	fi

	if [ -z "${FOUND_JARS}" ]; then
		return
	fi

	verbose "Checking all found jars and wars..." 2

	if [ -z "${UNZIP}" ]; then
		warn "unzip(1) not found, unable to peek into jars inside of jar!"
	fi
	for found in ${FOUND_JARS}; do
		pid="${found%%--*}"
		jar="${found#*--}"

		if [ -n "${UNZIP}" ]; then
			if zeroSize "${jar}"; then
				verbose "Skipping zero-size file '${jar}'..." 3
				continue
			fi
			jarjar="$(${UNZIP} -q -l "${jar}" 2>/dev/null | awk 'tolower($0) ~ /^ .*log4j.*[ejw]ar$/ { print $NF; }')"
			if [ -n "${jarjar}" ]; then
				extractAndInspect "${jar}" "${jarjar}" ${pid}
			fi
		fi

		checkInJar "${jar}" "${FATAL_CLASS}" "${pid}"
	done
}

checkOnlyGivenJars() {
	verbose "Checking only given jars..." 1
	FOUND_JARS="${CHECK_JARS}"
	checkJars
}

checkRpms() {
	verbose "Checking rpms..." 4

	local pkg version

	for pkg in $(rpm -qa --queryformat '%{NAME}--%{VERSION}\n' | grep log4j); do
		version="${pkg##*--}"
		if ! isFixedVersion "${version}"; then
			# Squeeze '--' so users don't get confused.
			pkg="$(echo "${pkg}" | tr -s -)"
			SUSPECT_PACKAGES="${SUSPECT_PACKAGES} ${pkg}"
		fi
	done
}

checkPackages() {
	if expr "${SKIP}" : ".*packages" >/dev/null; then
		verbose "Skipping package check." 2
		return
	fi

	verbose "Checking for vulnerable packages..." 2

	if [ x"$(command -v rpm 2>/dev/null)" != x"" ]; then
		checkRpms
	fi
}

checkPid() {
	local pid="${1}"
	verbose "Checking process ${pid} for command-line flags..." 6

	ps -www -q "${pid}" -o command= | grep -q -- "${FATAL_SETTING}"
}

checkProcesses() {
	local jars
	if expr "${SKIP}" : ".*processes" >/dev/null; then
		verbose "Skipping process check." 2
		return
	fi

	verbose "Checking running processes..." 3
	local lsof="$(command -v lsof 2>/dev/null || true)"
	if [ -z "${lsof}" ]; then
		jars="$(ps -o pid,command= -wwwax | awk 'tolower($0) ~ /[ejw]ar$/ { print $1 "--" $NF; }' | uniq)"
	else
		jars="$(${lsof} -c java 2>/dev/null | awk 'tolower($0) ~ /reg.*[ejw]ar$/ { print $2 "--" $NF; }' | uniq)"
	fi
	FOUND_JARS="${FOUND_JARS:+${FOUND_JARS} }${jars}"
}

cleanup() {
	if [ -n "${_TMPDIR}" ]; then
		rm -fr "${_TMPDIR}"
	fi
}

extractAndInspect() {
	local jar="${1}"
	local jarjar="${2}"
	local pid="${3}"
	local f

	verbose "Extracting ${jar} to look inside jars inside of jars..." 5

	cdtmp
	if ${UNZIP} -o -q "${jar}" ${jarjar} 2>/dev/null; then
	for f in ${jarjar}; do
		checkInJar "${f}" "${FATAL_CLASS}" ${pid} "${jar}"
	done
	fi
}

findJars() {
	verbose "Looking for jars..." 2
	checkProcesses
	checkFilesystem
}

isFixedVersion () {
	local version="${1}"
	local major minor

	major="${version%%.*}"  # 2.15.0 => 2
	minor="${version#*.}"   # 2.15.0 => 15.0
	tiny="${minor#*.}"     # 15.0 => 0

	# strip off any possible other sub-versions
	# e.g., 2.15.0.12345
	tiny="${tiny%%.*}"     # 0.12345 => 0
	minor="${minor%%.*}"   # 15.0 => 15

	# NaN => unknown
	if ! expr "${major}" : "[0-9]*$" >/dev/null; then
		return 1
	fi
	if ! expr "${minor}" : "[0-9]*$" >/dev/null; then
		return 1
	fi

	if [ ${major} -lt ${MAJOR_WANTED} ] ||
		[ ${major} -eq ${MAJOR_WANTED} -a ${minor} -lt ${MINOR_WANTED} ] ||
		[ ${major} -eq ${MAJOR_WANTED} -a ${minor} -eq ${MINOR_WANTED} -a ${tiny} -lt ${TINY_WANTED} ]; then
		SHOULD_UPGRADE="yes"
	fi

	if [ ${major} -lt ${MAJOR_WANTED} -o ${minor} -ge ${MINOR_MINIMUM} ]; then
		return 0
	fi

	return 1
}

log() {
	msg="${1}"
	echo "${LOGPREFIX}: ${msg}"
}

log4jcheck() {
	verbose "Running all checks..." 1

	checkPackages
	checkJars
}

usage() {
	cat <<EOH
Usage: ${PROGNAME} [-hv] [-j jar] [-s skip] [-p path]
	-h       print this help and exit
	-j jar   check only this jar
	-p path  limit filesystem traversal to this directory
	-s skip  skip these checks (files, packages, processes)
	-v       be verbose
EOH
}

verbose() {
	local readonly msg="${1}"
	local level="${2:-1}"
	local i=0

	if [ "${level}" -le "${VERBOSITY}" ]; then
		while [ ${i} -lt ${level} ]; do
			printf "=" >&2
			i=$(( ${i} + 1 ))
		done
		echo "> ${msg}" >&2
	fi
}

verdict() {
	if [ -z "${SUSPECT_JARS}" -a -z "${SUSPECT_PACKAGES}" -a -z "${SUSPECT_CLASSES}" ]; then
		log "No obvious indicators of vulnerability to CVE-2021-44228 / CVE-2021-45046 found."
		RETVAL=0
	fi

	if [ -n "${SUSPECT_JARS}" ]; then
		echo
		echo "The following archives were found to include '${FATAL_CLASS}':"
		echo "${SUSPECT_JARS# *}" | tr ' ' '\n'
	fi

	if [ -n "${SUSPECT_PACKAGES}" ]; then
		echo
		echo "The following packages might still be vulnerable:"
		echo "${SUSPECT_PACKAGES}"
		RETVAL=1
	fi

	if [ x"${SHOULD_UPGRADE}" = x"yes" ]; then
		echo
		echo "Note: You appear to be using (at least some version of) log4j <= ${MAJOR_WANTED}.${MINOR_WANTED}.${TINY_WANTED}."
		echo "You should upgrade to that or a later version even if no obvious"
		echo "vulnerability to CVE-2021-44228 / CVE-2021-45046 was reported."

		if [ ${RETVAL} -eq 0 ]; then
			RETVAL=2
		fi
	fi
}

warn() {
	msg="${1}"
	echo "${LOGPREFIX}: ${msg}" >&2
}

zeroSize() {
	local file="${1}"
	local size

	# stat(1) is not portable :-/
	size="$(ls -l "${file}" | awk '{print $5}')"
	if [ x"${size}" = x"0" ]; then
		return 0
	fi

	return 1
}

###
### Main
###

trap 'cleanup' 0

while getopts 'Vhj:s:p:v' opt; do
	case "${opt}" in
		V)
			echo "${PROGNAME} ${VERSION}"
			exit 0
			# NOTREACHED
		;;
		h\?)
			usage
			exit 0
			# NOTREACHED
		;;
		j)
			d="${OPTARG%/*}"
			if [ x"${d}" = x"${OPTARG}" ]; then
				d="."
			fi
			f="$(cd "${d}" && pwd)/${OPTARG##*/}"
			CHECK_JARS="${CHECK_JARS:+${CHECK_JARS} }${f}"
		;;
		p)
			SEARCH_PATHS="${SEARCH_PATHS:+${SEARCH_PATHS} }$(cd "${OPTARG}" && pwd)/."
		;;
		s)
			SKIP="${SKIP} ${OPTARG}"
		;;
		v)
			VERBOSITY=$(( ${VERBOSITY} + 1 ))
		;;
		*)
			usage
			exit 1
			# NOTREACHED
		;;
	esac
done
shift $(($OPTIND - 1))

if [ $# -gt 0 ]; then
	usage
	exit 1
	# NOTREACHED
fi

if [ -z "${CHECK_JARS}" ]; then
	log4jcheck
else
	checkOnlyGivenJars
fi
verdict

exit ${RETVAL}
